airflow/jobs/local_task_job.py

```diff
@@ -98,7 +98,10 @@ def signal_handler(signum, frame):
 
             heartbeat_time_limit = conf.getint('scheduler', 'scheduler_zombie_task_threshold')
 
-            while True:
+            # task callback invocation happens either here or in
+            # self.heartbeat() instead of taskinstance._run_raw_task to
+            # avoid race conditions
+            while not self.terminating:
```

Comment:
What's this change for? How come we didn't do it before?

Reply:
This is needed because of `self.handle_task_exit(return_code)`. If we restart the loop after heartbeat sets `self.terminating` to True, then handle_task_exit will be invoked which will result in duplicated callback invocations. I added more comment to the code to document this.